src/main/scala/com/microsoft/hyperspace/index/sources/delta/DeltaLakeRelation.scala

```diff
@@ -142,4 +143,95 @@ class DeltaLakeRelation(spark: SparkSession, override val plan: LogicalRelation)
       /* modificationTime */ modificationTime,
       path)
   }
+
+  private def deltaLakeVersionHistory(index: IndexLogEntry): Seq[(Int, Int)] = {
+    val versions =
+      index.derivedDataset.properties.properties
+        .getOrElse(DeltaLakeConstants.DELTA_VERSION_HISTORY_PROPERTY, "")
+
+    // The value is comma separated versions - <index log version>:<delta table version>.
+    // e.g. "1:2,3:5,5:9"
+    versions.split(",").reverseIterator.foldLeft(Seq[(Int, Int)]()) { (list, versionStr) =>
+      val pair = versionStr.split(":")
+      val Array(indexLogVersion, deltaTableVersion) = versionStr.split(":").map(_.toInt)
+      if (list.nonEmpty && list.head._2 == deltaTableVersion) {
+        // Omit if the delta lake version is the same. In this case, we only take the higher
+        // index log version as this is usually from index optimizations.
+        list
+      } else {
+        (indexLogVersion, deltaTableVersion) +: list
+      }
+    }
+  }
+
+  /**
+   * Returns IndexLogEntry of the closest index version for the given relation.
+   *
+   * curFiles is used to calculate the similarity with each index version data.
+   * Delta Lake source provider utilizes the history information of DELTA_VERSION_HISTORY_PROPERTY
+   * to find the closet version of the index.
+   *
+   * @param index Candidate index to be applied.
+   * @return IndexLogEntry of the closest version among available index versions.
+   */
+  override def closestIndexVersion(index: IndexLogEntry): IndexLogEntry = {
+    // Seq of (index log version, delta lake table version)
+    val versions = deltaLakeVersionHistory(index)
+
+    lazy val indexManager = Hyperspace.getContext(spark).indexCollectionManager
+    def getIndexLogEntry(logVersion: Int): IndexLogEntry = {
+      indexManager
+        .getIndexLogEntry(index.name, logVersion)
+        .get
+        .asInstanceOf[IndexLogEntry]
+    }
+    // TODO: Currently assume all versions of index data exist.
+    //  Need to check and remove candidate indexes.
+
+    plan.relation match {
+      case HadoopFsRelation(location: TahoeLogFileIndex, _, _, _, _, _) =>
+        val equalOrLessLastIndex = versions.lastIndexWhere(location.tableVersion >= _._2)
```

Comment:
equalOrLessThanLastIndex? Bt the way, could you add a comment what this variable actually means (e.g, what is this index supposed to point to)?